x-pack/plugin/inference/src/internalClusterTest/java/org/elasticsearch/xpack/inference/action/filter/ShardBulkInferenceActionFilterIT.java

```diff
@@ -0,0 +1,198 @@
+                    }
+                }
+            }
+            """).setSettings(shardsSettings).get();
+
+        int totalBulkReqs = randomIntBetween(2, 100);
+        long totalDocs = 0;
+        for (int bulkReqs = 0; bulkReqs < totalBulkReqs; bulkReqs++) {
+            BulkRequestBuilder bulkReqBuilder = client().prepareBulk();
+            int totalBulkSize = randomIntBetween(1, 100);
+            for (int bulkSize = 0; bulkSize < totalBulkSize; bulkSize++) {
+                String id = Long.toString(totalDocs);
+                boolean isIndexRequest = randomBoolean();
+                Map<String, Object> source = new HashMap<>();
+                source.put("sparse_field", isIndexRequest && rarely() ? null : randomAlphaOfLengthBetween(0, 1000));
+                source.put("dense_field", isIndexRequest && rarely() ? null : randomAlphaOfLengthBetween(0, 1000));
+                if (isIndexRequest) {
+                    bulkReqBuilder.add(new IndexRequestBuilder(client()).setIndex(INDEX_NAME).setId(id).setSource(source));
+                    totalDocs++;
+                } else {
+                    boolean isUpsert = randomBoolean();
+                    UpdateRequestBuilder request = new UpdateRequestBuilder(client()).setIndex(INDEX_NAME).setDoc(source);
+                    if (isUpsert || totalDocs == 0) {
+                        request.setDocAsUpsert(true);
+                        totalDocs++;
+                    } else {
+                        // Update already existing document
+                        id = Long.toString(randomLongBetween(0, totalDocs - 1));
+                    }
+                    request.setId(id);
+                    bulkReqBuilder.add(request);
+                }
+            }
+            BulkResponse bulkResponse = bulkReqBuilder.get();
+            if (bulkResponse.hasFailures()) {
+                // Get more details in case something fails
+                for (BulkItemResponse bulkItemResponse : bulkResponse.getItems()) {
+                    if (bulkItemResponse.isFailed()) {
+                        fail(
+                            bulkItemResponse.getFailure().getCause(),
+                            "Failed to index document %s: %s",
+                            bulkItemResponse.getId(),
+                            bulkItemResponse.getFailureMessage()
+                        );
+                    }
+                }
+            }
+            assertFalse(bulkResponse.hasFailures());
+        }
+
+        client().admin().indices().refresh(new RefreshRequest(INDEX_NAME)).get();
+
+        SearchSourceBuilder sourceBuilder = new SearchSourceBuilder().size(0).trackTotalHits(true);
+        SearchResponse searchResponse = client().search(new SearchRequest(INDEX_NAME).source(sourceBuilder)).get();
+        assertThat(searchResponse.getHits().getTotalHits().value, equalTo(totalDocs));
+        searchResponse.decRef();
```

Comment:
Should this be in a try-finally block so that `searchResponse.decRef()` is always called?

Reply:
Good point, changed! 36927cf59fef9e53eb375c58f723c8dc3711d314